### ISSUE TITLE
Remove `strlen` as a "low performing construction" when used in a loop

### ISSUE THREAD

**Issue by zerkms** [kalessil/phpinspectionsea#1673]
Description (screenshot):
--

Example
--

```php
for ($i = 0; $i < strlen($prj); ++$i) {}
```
Here `strlen` is marked as "low performing construction". While here
```php
for ($i = 0; $i < foo($prj); ++$i) {}
```
`foo` is not marked as such. The truth is - `strlen` is heavily optimised and is practically a single memory address (string structure field) load. From that perspective - `foo` would be significantly worse performing, than `strlen`, yet, only the latter is marked as "slow".

Is there any chance cargo cult of "`count` and `strlen` are slow" is finally ended?

**Comment by orklah** [comment]
Unless foo() can be flagged as Pure (it may be an evolution based on new attributes by Jetbrains), it is impossible to be sure it returns the same result each time. Thus, it can't be initialized in the first part of the for.

This is not true for strlen or count. Even if they're highly optimised, the for will still call the count opcode as many times as the for iterates. The suggestion will make it call once.

This may be a micro-optimisation, we could even argue that the extra assignment will make it less quick for empty or small enough arrays, but the suggestion is correct

**Comment by zerkms** [comment]
@orklah I can see it as micro-optimisation, but claiming at as "low performing" or "cause performance issues when counting big arrays" is simply _wrong_.

**Comment by orklah** [comment]
The inspection looks for array modification and will not suggest the change if it sees it.

I'm not sure what your point is. There is no arguing that calling count() once is quicker than calling it twice. If the whole discussion is about scale, then yes, it's probably meaningless on the scale of a big application. On the other hand, for hot paths, it may gain a significant time

**Comment by zerkms** [comment]
> I'm not sure what your point is

My point is: it should be better wording. Neither of them is slow, so should not be called "low performing" or "cause performance issues when counting big arrays". It's a lot of iterations case "performance issues", not one memory address load.

**Comment by orklah** [comment]
Yeah probably, the wording could be better, I think you could propose something, it should be accepted. 

On the other hand, if this inspection could take advantage of [#Pure] , the wording would be on point because we don't know what foo() could contain. In any case, it's better practice to initialize the variables in the first part of the for, no matter how performant the function is 😄

**Comment by zerkms** [comment]
I measured - on my machine `count` every loop iteration adds ~1ms for every 100k elements.

**Comment by ea-inspections-team** [comment]
Thanks for reporting @zerkms and @orklah for joining the discussion!

@kalessil, mind sharing your opinion here?

**Comment by rentalhost** [comment]
I did a [**benchmark**](https://pastebin.com/nJEFtt3x) with 64k string. The results for PHP 8 on my setup are:

* **variable with static strlen()**: reference;
* **for($j = max; $i < $j)**: no change;
* **for(; < strlen())**: +11~12% slow;
* **variable with static foo()**: + 1% slow (technical tie);
* **for(; < foo())**: +880% slow;

So, in fact, strlen() is very well optimized and could be considered as a micro-optimization.

For the foo() case, I think that it must be moved to outside, except if it could change de code behaviour. But in this case, I must consider as exception and a `@noinspection` must be set.

**Comment by zerkms** [comment]
@rentalhost thanks!

Here are my results

```
Case for < $j = strlen()           duration of 8.159062s (avg. each 0.00271968s) +0.00%
Case variable with static strlen() duration of 8.164859s (avg. each 0.00272161s) +0.07%
Case variable with static foo()    duration of 8.282229s (avg. each 0.00276074s) +1.51%
Case for < strlen()                duration of 9.431748s (avg. each 0.00314391s) +15.60%
Case for < foo()                   duration of 57.166652s (avg. each 0.01905555s) +600.65%
```

`strlen()` in a loop indeed adds 15.6% of relative latency, which is still just 1ms per loop.

**Comment by kalessil** [comment]
@ea-inspections-team: this inspection definitely needs some love - it's plenty of false-positives (inspect Symfony for more details).

On the subject: strlen and count are indeed well optimized in the latest PHP versions (complexity-wise and opcodes-wise).
This is one of the earliest inspections and I've started the project in 2014, so it's a good to point to revisit this part.

- Message: good points in the discussion from @zerkms 
- A good point about the proposal about pure function from @orklah 
- Based on benchmarks from @rentalhost  and @zerkms, I'd imagine it's rather repetitive calls (control flow optimizations).

What I'm not sure about is if should be a new inspection, or a new setting like "Suggest optimizing loops conditions". For me personally, the both ways will work and the inspection was of great value since it was introduced.
